UGENE-7432. Implement GUI tests

evelinabis opened the pull request.

Commit:
UGENE-7432. Implement GUI tests.
First partial implementation.

Commit:
Merge branch 'pcr_primer_design_for_dna_assembly' of https://github.com/ugeneunipro/ugene into UGENE-7432

Commit:
UGENE-7432. Implement GUI tests.

1. Increased the width of the PCR Primer Design tab.
2. Increased the width of the Color column in the Annotations
   Highlighting tab.
3. Corrected tests related to changes 1-2.
4. Refactored PCR Primer Design utils.
5. Refactored PCR Primer Design tests.
6. Added missing PCR Primer Design tests.

Commit:
UGENE-7432. Implement GUI tests.
Build fixed.

yalgaer commented on src/corelibs/U2Gui/src/options_panel/GroupOptionsWidget.h:
```diff
@@ -69,7 +69,7 @@ class U2GUI_EXPORT GroupOptionsWidget : public QWidget {
 
     static const int TITLE_HEIGHT = 30;
-    static const int MIN_WIDGET_WIDTH = 440;
+    static const int MIN_WIDGET_WIDTH = 480;  // Minimum for the tab of sequence viewer "PCR Primer Design".
```
This will change width for all tabs in all views (assembly, sequence, msa, mca, tree)  regardless if pcr plugin is loaded or not.
This should be a dynamic per-tab variable set by the plugin.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsPcrPrimerDesignForDnaAssembly.h:
```diff
@@ -0,0 +1,173 @@
+        SearchArea(U2Range<int> r, bool selectManually = false);
+
+        U2Range<int> region;
+        bool         selectManually = false;
```
Please reformat the file with the clangformat style

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsPcrPrimerDesignForDnaAssembly.h:
```diff
@@ -0,0 +1,173 @@
+    static void openTab(GUITestOpStatus& os);
+
+    // Sets forward (if direction=Direct) or reverse (if direction=Complementary) user primer.
+    static void setUserPrimer(GUITestOpStatus& os, const QString& primer, U2Strand::Direction direction);
```
Use const & for all non-standard types (U2Strand::Direction)

Commit:
UGENE-7432. Implement GUI tests

Minor changes.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsPcrPrimerDesignForDnaAssembly.cpp:
```diff
@@ -0,0 +1,367 @@
+    };
+    const auto moveMouseAndScroll = [](QWidget* moveMouseTo, bool scrollUp) {
+        GTMouseDriver::moveTo(GTWidget::getWidgetCenter(moveMouseTo));
+        const int scrollCoef = 20;
```
Dont use consts for locals (you do not use them consistently in this CL and UGENE does not use them across the code too)

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsPcrPrimerDesignForDnaAssembly.cpp:
```diff
@@ -0,0 +1,367 @@
+        return;  // Nothing to do.
+    }
+
+    tab = (tab == nullptr ? getTabWgt(os) : tab);
```
nit: braces are not needed

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsPcrPrimerDesignForDnaAssembly.cpp:
```diff
@@ -0,0 +1,367 @@
+static bool isTableIndexCorrect(GUITestOpStatus& os, int ind, const QTableWidget* table) {
+    GT_CHECK_RESULT(table != nullptr, "Table is nullptr", false)
+    const int tableRowCount = table->rowCount();
+    GT_CHECK_RESULT(0 <= ind && ind < tableRowCount, QString("Invalid index of '%1': expected 0<=i<%2, current i=%3").
```
Don't use Yoda style. Write code in the same order as you read it.

DmitriiSukhomlinov replied:
I disagree. This kind of the order texting is much more clear and understandable. This is kinda a ray, the largest numbers are righter. The analogue of this sign:

`0 <= ind < tableRowCount`

This king of style makes it easy to imagine a picture like this:
![image](https://user-images.githubusercontent.com/26257527/136163375-d46cabec-6bf2-466a-9a2e-a93af757c500.png)

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsPcrPrimerDesignForDnaAssembly.cpp:
```diff
@@ -0,0 +1,367 @@
+ */
+static void setRange(GUITestOpStatus& os, const QString& minSbName, const QString& maxSbName, U2Range<int> range,
+                     const QWidget* parent) {
+    if (QSpinBox* min = GTWidget::findSpinBox(os, minSbName, parent)) {
```
min is never nullptr here

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsPcrPrimerDesignForDnaAssembly.cpp:
```diff
@@ -0,0 +1,367 @@
+void GTUtilsPcrPrimerDesign::setParametersToExcludeInWholePrimers(GUITestOpStatus& os,
+                                                                  ParametersToExcludeInWholePrimers params) {
+    const QWidget* tab = getTabWgt(os);
+    QWidget* const parent = GTWidget::findWidget(os, "wgtParameters2ExcludeInWholePrimers", tab);
```
remove const  "spam" from this line

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsPcrPrimerDesignForDnaAssembly.cpp:
```diff
@@ -0,0 +1,367 @@
+}
+
+// Return the table of result sequences. Doesn't open the tab. Returns nullptr in case of an error.
+static QTableWidget* getResultTable(GUITestOpStatus& os, const QWidget* tab) {
```
In case of an error sets the error into "os" or some other state holder structure. Never return nullptr from getters like in all other selector methods. Use "flags" to adjust the behaviour

evelinabis replied:
> Never return nullptr from getters like in all other selector methods.

And what to return if widget is not found?

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsPcrPrimerDesignForDnaAssembly.cpp:
```diff
@@ -0,0 +1,367 @@
+    const QWidget* tab = getTabWgt(os);
+    const QWidget* parent = GTWidget::findWidget(os, "wgtParametersOfPrimingSequences", tab);
+    scrollToWidget(os, parent, tab);
+    setRange(os, "sbMinRequireGibbs",         "sbMaxRequireGibbs",         params.gibbsFreeEnergy, parent);
```
Reformat the file according to the clangformat

Commit:
UGENE-7432. Implement GUI tests.

Changes by review.
Return to the previous state of the options panel. Return covers the
tests too.
Adding line breaks at the ends of files.

Commit:
UGENE-7432. Implement GUI tests.

Minor readability changes.

yalgaer commented on src/corelibs/U2Gui/src/options_panel/OptionsPanelWidget.cpp:
```diff
@@ -42,12 +42,17 @@ OptionsScrollArea::OptionsScrollArea(QWidget *parent)
                   " }");
     setSizePolicy(QSizePolicy::Minimum, QSizePolicy::Preferred);
     setMaximumWidth(GroupOptionsWidget::getMaxWidgetWidth());
+    setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
```
Do we really need this?

If a user has a large font set by default all our tabs won't fit to the pre-defined tab width and will not be accessible for a such user at all.

You can reproduce this problem by setting "Text Scaling Factor" to 2x on Linux desktop.

yalgaer commented on src/corelibs/U2Gui/src/options_panel/OptionsPanelWidget.cpp:
```diff
@@ -164,6 +169,9 @@ GroupOptionsWidget *OptionsPanelWidget::createOptionsWidget(const QString &group
 
     optionsWidgets.insert(0, groupWidget);
 
+    optionsScrollArea->setMinWidth(mainWidget->minimumWidth() + 20);
```
nit: add a comment what is '20'

yalgaer commented on src/corelibs/U2Gui/src/options_panel/OptionsPanelWidget.h:
```diff
@@ -42,6 +42,12 @@ class OptionsScrollArea : public QScrollArea {
     /** Ensures that the scroll area would have an appropriate width */
     virtual QSize sizeHint() const;
+
+    /** Set the minimum width of a child. */
```
nit: Set -> Sets

yalgaer replied:
QWidget already has "setMinimumWidth" and "setMinWidth" sounds ambiguous. Could you please name it more explicitly, like setMinimumWidthHint or similar.

yalgaer reviewed (changes requested):
Looks like a patch for the minWidth is very small. I can review it as a part of the current PR but would suggest forking it into an independent patch targeted directly to master.

In this case as an option panel test candidate for a different size we can select any existing tab that can benefit from size change.

Commit:
UGENE-7432. Implement GUI tests.

Removed changes related to increasing Option Panel width.

Commit:
UGENE-7432. Implement GUI tests.

Reformat with the clangformat style.

Commit:
UGENE-7432. Implement GUI tests.

Removed consts for locals (by review).

Commit:
UGENE-7432. Implement GUI tests.

Comments indentation is unified.

evelinabis commented:
#601 #605
